File: PC-PLANET-BACKEND/src/main/java/com/pcplanet/entity/Product.java
package com.pcplanet.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.pcplanet.enums.ProductStatus;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.Setter;

import java.util.List;

@Getter
@Setter
@Entity
@Table(name = "product")
public class Product extends BaseEntity {

    @NotBlank
    @Column(name = "name")
    private String name;

    @NotBlank
    @Column(name = "code")
    @Size(min = 6, max = 12)
    private String code;

    @NotBlank
    @Column(name = "model")
    private String model;

    @NotNull
    @Column(name = "price")
    private Double price;

    @Column(name = "warranty")
    private Integer warranty;

    @NotNull
    @Column(name = "status", nullable = false)
    @Enumerated(EnumType.STRING)
    private ProductStatus status;

    @ManyToOne(optional = false)
    @JoinColumn(name = "brand_id", foreignKey = @ForeignKey(name = "fk_product_brand"))
    private Brand brand;

    @ManyToOne(optional = false)
    @JoinColumn(name = "category_id", foreignKey = @ForeignKey(name = "fk_product_category"))
    private Category category;

    @ManyToOne
    @JoinColumn(name = "sub_category_id", foreignKey = @ForeignKey(name = "fk_product_sub_category"))
    private SubCategory subCategory;

    @JsonIgnore
    @OneToMany(mappedBy = "product", cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE}, orphanRemoval = true)
    private List<ProductKeyFeature> keyFeatures;

    @JsonIgnore
    @OneToMany(mappedBy = "product", cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE}, orphanRemoval = true)
    private List<ProductSpecification> specifications;

    @JsonIgnore
    @OneToMany(mappedBy = "product", cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE}, orphanRemoval = true)
    private List<ProductDescription> descriptions;

    @JsonIgnore
    @OneToMany(mappedBy = "product", cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE}, orphanRemoval = true)
    @OrderBy("id ASC")
    private List<ProductImage> images;

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
            name = "product_product_attribute_value",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "attribute_value_id")
    )
    private List<ProductAttributeValue> attributeValues;
}

File: pc-planet-frontend/src/features/models/Product.ts
export type ProductStatus = 'IN_STOCK' | 'OUT_OF_STOCK' | 'UP_COMING' | 'PRE_ORDER';

export type ProductBrand = {
  id: number | null;
  name?: string;
};

export type ProductCategory = {
  id: number | null;
  name?: string;
  brands?: ProductBrand[];
};

export type ProductSubCategory = {
  id: number | null;
  name?: string;
};

export type ProductKeyFeature = {
  id: number | null;
  name: string;
  value: string;
};

export interface ProductInfo {
  id: number | null;
  name: string;
  code: string;
  model: string;
  price: number | null;
  status: ProductStatus | undefined;
  brand: ProductBrand | null;
  category: ProductCategory | null;
  keyFeatures?: ProductKeyFeature[];
  image: string;
}

export type PropertyValue = {
  id: number | null;
  value: string;
};

export type SpecificationProperty = {
  id: number | null;
  name: string;
  propertyValues: PropertyValue[];
};

export type ProductSpecification = {
  id: number | null;
  type: string;
  properties: SpecificationProperty[];
};

export type ProductDescription = {
  id: number | null;
  name: string;
  value: string;
};

export type ProductAttributeValue = {
  id: number | null;
  value?: string;
};

export type ProductAttribute = {
  id: number | null;
  name: string;
  attributeValues?: ProductAttributeValue[];
};

export type ProductImage = {
  id: number | null;
  fileName: string;
};

export interface ProductDetailsModel extends ProductInfo {
  warranty?: number;
  specifications?: ProductSpecification[];
  descriptions?: ProductDescription[];
  attributeValues?: ProductAttributeValue[];
  images?: ProductImage[];
}
